fusion/FSL.md

```diff
@@ -0,0 +1,1206 @@
+information is passed to the constraints clause. On a match the sequence
+clause returns \_pass\_ to the Fusion API.
+
+The index field is the zero-referenced position of the start of the match 
+within ioput. The length is the number of instructions, inclusive of the
+starting instruction.
+
+An instruction sequence is expressed as a simple list of Mavis-assigned
+unique identifiers, or as a set of assembly language instructions with
+abstracted operands. The choice is based on the constraints.
+
+The UID list is sufficient if there are no constraints on the operands
+in the list of instructions in the sequence.
+
+For example, these two instructions are a frequent sequence in compiled
+C.  The left column is the encoding, the comment is the Mavis unique identifier 
+for each instruction which is independent of the operands.
+
+```
+0x0512     c.slli x10,4    ; 0xf
+0x8111     c.srli x10,4    ; 0x13
+```
+
+If you do not require constraints on the operands and therefore have no
+need to refer to the operands by name, you can simply specify the UIDs 
+as a list, one per line.
+
+```
+ioput iop1
+
+sequence seq1(iop1,rv64g) {
+  0xf
+  0x13
+}
+```
+
+With the UID sequence expression there are no operand specific 
+constraints and therefore the conversion clause must handle any legal
+combinations of the two rd and two constants.
+
+This results in a transform to a generalized fusion op of a shift left 
+followed by right and not the zero extension expressed by x10,4.
+
+If the constraints clause implements operand contraints the instruction
+sequence should be expressed using the abstract assembly syntax so the
+operands can be referenced by name.
+
+The expression below takes advantage of positional constraints:
+
+```
+ioput iop1
+
+sequence seq1(iop1,rv64g) {
+  c.slli g1,c1
+  c.srli g1,c1
+}
+```
+
+In this case the operands labeled g1 are, by association, required to be
+the same value, similarly with the constant operands.
+
+Another option would be to fully enumerate the operands and let the
+constraints clause enforce the required limits on operands.
+
+```
+ioput iop1
+
+sequence seq1(iop1,rv64g) {
+  c.slli g1,c1
+  c.srli g2,c2
+}
+```
+
+Note both the later two cases can be made equivalent by construction of
+the equivalent constraints in the clause.
+
+Now in both cases the transformation expresses a fusion operation for
+zero extention of the common rd register by the common constant.
+
+```
+g1 = g2
+c1 = c2
+```
+
+In some cases positional constraints can save development time at no 
+expense to performance.
+
+The example sequences above have an implied strict ordering. There are 
+sequence pragmas to relax the strictness of sequence matching. 
+
+The required pragma, \_req\_, indicates that an unspecified instruction
+is required in that position. The \_req\_ case does not constrain 
+what instruction can be present in the gap.
+
+The optional pragma, \_opt\_, indicates that the match will not be 
```

Comment:
Is there a need for this opt instruction to possibly be a variable number of instructions (e.g. 1-3 optional instructions)

Reply:
I will add an integer to the syntax, __opt__ 3, meaning any of 1,2,3 optional instructions. 

Implication is that no instructions in the range is also legal, since each optional instruction is 0 or 1.

in the syntax __opt__ will be treated the same as __opt__ 1

Makes sense to do the same for __req__ 3

Thanks.

Reply:
@rajatbhatia1 I have made these changes they are included in [PR #163](https://github.com/riscv-software-src/riscv-perf-model/pull/163). I am going to close this PR it is redundant now.
```
The required pragma, \_req\_, indicates that an unspecified instruction
is required in that position. The \_req\_ case does not constrain
what instruction can be present in the gap. A trailing integer
can specify more than one required instruction.

The optional pragma, \_opt\_, indicates that the match will not be
rejected if there is an instruction or not in this position. A
trailing integer can be used to specify up to N optional instructions.

_opt_ 2    indicates 0 - 2 instructions will match
_req_ 2    indicates there must be 2 instructions in this position
           in the sequence.
```